internal/component/prometheus/remote/queue/filequeue/filequeue_test.go

```diff
@@ -0,0 +1,248 @@
+package filequeue
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/vladopajic/go-actor/actor"
+	"go.uber.org/goleak"
+
+	"github.com/grafana/alloy/internal/component/prometheus/remote/queue/types"
+
+	"github.com/go-kit/log"
+	"github.com/stretchr/testify/require"
+)
+
+func TestFileQueue(t *testing.T) {
+	defer goleak.VerifyNone(t)
+	dir := t.TempDir()
+	log := log.NewNopLogger()
+	mbx := actor.NewMailbox[types.DataHandle]()
+	mbx.Start()
+	defer mbx.Stop()
+	q, err := NewQueue(dir, func(ctx context.Context, dh types.DataHandle) {
+		_ = mbx.Send(ctx, dh)
+	}, log)
+	require.NoError(t, err)
+	q.Start()
+	defer q.Stop()
+	err = q.Store(context.Background(), nil, []byte("test"))
+
+	require.NoError(t, err)
+
+	meta, buf, err := getHandle(t, mbx)
+	require.NoError(t, err)
+	require.True(t, string(buf) == "test")
+	require.Len(t, meta, 0)
+
+	// Ensure nothing new comes through.
+	timer := time.NewTicker(100 * time.Millisecond)
+	select {
+	case <-timer.C:
+		return
+	case <-mbx.ReceiveC():
+		require.True(t, false)
+	}
+}
+
+func TestMetaFileQueue(t *testing.T) {
+	defer goleak.VerifyNone(t)
+
+	dir := t.TempDir()
+	log := log.NewNopLogger()
+	mbx := actor.NewMailbox[types.DataHandle]()
+	mbx.Start()
+	defer mbx.Stop()
+	q, err := NewQueue(dir, func(ctx context.Context, dh types.DataHandle) {
+		_ = mbx.Send(ctx, dh)
+	}, log)
+	q.Start()
+	defer q.Stop()
+	require.NoError(t, err)
+	err = q.Store(context.Background(), map[string]string{"name": "bob"}, []byte("test"))
+	require.NoError(t, err)
+
+	meta, buf, err := getHandle(t, mbx)
+	require.NoError(t, err)
+	require.True(t, string(buf) == "test")
+	require.Len(t, meta, 1)
+	require.True(t, meta["name"] == "bob")
+}
+
+func TestCorruption(t *testing.T) {
+	defer goleak.VerifyNone(t)
+
+	dir := t.TempDir()
+	log := log.NewNopLogger()
+	mbx := actor.NewMailbox[types.DataHandle]()
+	mbx.Start()
+	defer mbx.Stop()
+	q, err := NewQueue(dir, func(ctx context.Context, dh types.DataHandle) {
+		_ = mbx.Send(ctx, dh)
+	}, log)
+	q.Start()
+	defer q.Stop()
+	require.NoError(t, err)
+
+	err = q.Store(context.Background(), map[string]string{"name": "bob"}, []byte("first"))
+	require.NoError(t, err)
+	err = q.Store(context.Background(), map[string]string{"name": "bob"}, []byte("second"))
+
+	require.NoError(t, err)
+
+	// Send is async so may need to wait a bit for it happen.
+	require.Eventually(t, func() bool {
```

Comment:
Could this be flaky? when Eventually doesn't catch the 1.committed existing, cause it appears and disappears too fast?

Would be good to run this test with `-count 100` or sth to validate it's not flaky if you didn't already :)

Reply:
I just realised that we won't delete the file until we call `Get` on the data handle... not sure if I like this side-effect, let's discuss in another thread.